File: Claude/Simples/dfs/dfs.py
#Input: Using the previous graph, how to implement the Depth-First Search algorith?

from collections import deque

class Graph:
    def __init__(self):
        self.adjacency_list = {}

    def add_vertex(self, vertex):
        if vertex not in self.adjacency_list:
            self.adjacency_list[vertex] = []

    def add_edge(self, vertex1, vertex2):
        if vertex1 in self.adjacency_list and vertex2 in self.adjacency_list:
            self.adjacency_list[vertex1].append(vertex2)
            self.adjacency_list[vertex2].append(vertex1)

    def print_graph(self):
        for vertex in self.adjacency_list:
            print(vertex, ":", self.adjacency_list[vertex])

    def dfs(self, start_vertex):
        visited = set()
        stack = deque([start_vertex])

        while stack:
            vertex = stack.pop()
            if vertex not in visited:
                visited.add(vertex)
                print(vertex)
                for neighbor in self.adjacency_list[vertex]:
                    stack.append(neighbor)

# Example usage
graph = Graph()
graph.add_vertex('A')
graph.add_vertex('B')
graph.add_vertex('C')
graph.add_vertex('D')
graph.add_vertex('E')

graph.add_edge('A', 'B')
graph.add_edge('A', 'C')
graph.add_edge('B', 'D')
graph.add_edge('B', 'E')

print("Depth-First Search (DFS) traversal:")
graph.dfs('A')

"""
Output: Depth-First Search (DFS) traversal:
        A
        C
        B
        E
        D
Wrong:
Should be: Depth-First Search (DFS) traversal:
        A
        B
        D
        E
        C
Score: 0.5
"""

File: Claude/Simples/dfs/Dfs.cpp
// Input: Using the previous graph, how to implement the Depth-First Search algorith?

#include <iostream>
#include <vector>
#include <list>
#include <stack>

// Graph class
class Graph {
private:
    int numVertices;
    std::vector<std::list<int>> adjacencyList;
    std::vector<bool> visited;

public:
    // Constructor
    Graph(int vertices) {
        numVertices = vertices;
        adjacencyList.resize(vertices);
        visited.resize(vertices, false);
    }

    // Add an edge
    void addEdge(int src, int dest) {
        adjacencyList[src].push_back(dest);
        adjacencyList[dest].push_back(src); // Undirected graph
    }

    // Depth-First Search
    void dfs(int startVertex) {
        std::stack<int> stack;
        visited[startVertex] = true;
        stack.push(startVertex);

        while (!stack.empty()) {
            int currentVertex = stack.top();
            std::cout << currentVertex << " ";
            stack.pop();

            for (int neighbor : adjacencyList[currentVertex]) {
                if (!visited[neighbor]) {
                    visited[neighbor] = true;
                    stack.push(neighbor);
                }
            }
        }
    }

    // Print the graph
    void printGraph() {
        for (int i = 0; i < numVertices; ++i) {
            std::cout << "Vertex " << i << ":";
            for (int neighbor : adjacencyList[i]) {
                std::cout << " -> " << neighbor;
            }
            std::cout << std::endl;
        }
    }
};

int main() {
    Graph g(5);

    g.addEdge(0, 1);
    g.addEdge(1, 3);
    g.addEdge(1, 2);
    g.addEdge(0, 4);

    std::cout << "\nDepth-First Search (starting from vertex 0):" << std::endl;
    g.dfs(0);

    return 0;
}
/*
Output:
    Depth-First Search (starting from vertex 0):
    0 4 1 2 3

Wrong:
Should be:
    Depth-First Search (starting from vertex 0):
    0 1 3 2 4
Score: 0.5
*/